File: fabtrk/fabtrk/report/ft_drawings/ft_drawings.py
# # Copyright (c) 2026, UpGo Technologies and contributors
# # For license information, please see license.txt

import frappe

def execute(filters=None):
    filters = filters or {}

    columns = [
        {"label": "Project",        "fieldname": "project_name",  "fieldtype": "Link",  "options": "FT Project", "width": 190},
        {"label": "Drawing Number", "fieldname": "drawing_number", "fieldtype": "Data",  "width": 200, "align": "center"},
        {"label": "PO Serial No",   "fieldname": "po_serial_no",  "fieldtype": "Data",  "width": 190, "align": "center"},
        {"label": "Unit Weight",    "fieldname": "unit_weight",   "fieldtype": "Float", "width": 200, "align": "center"},
        {"label": "Required Qty",   "fieldname": "quantity",      "fieldtype": "Int",   "width": 190, "align": "center"},
        {"label": "Total Weight",   "fieldname": "total_weight",  "fieldtype": "Float", "width": 200, "align": "center"},
    ]

    conditions = ""
    values = {}

    if filters.get("project_number"):
        conditions += " AND p.name IN %(project_number)s"
        values["project_number"] = tuple(filters.get("project_number"))

    if filters.get("drawing_number"):
        conditions += " AND ad.drawing_number IN %(drawing_number)s"
        values["drawing_number"] = tuple(filters.get("drawing_number"))

    if filters.get("po_serial_no"):
        conditions += " AND pod.po_serial_no IN %(po_serial_no)s"
        values["po_serial_no"] = tuple(str(x) for x in filters.get("po_serial_no"))

    if filters.get("is_active"):
        conditions += " AND p.is_active = 1"

    query = f"""
        SELECT
            p.name AS project_name,
            ad.name AS drawing_id,
            ad.drawing_number AS drawing_number,
            pod.po_serial_no AS po_serial_no,
            IFNULL(ad.unit_weight, 0) AS unit_weight,
            IFNULL(pod.required_qty, 0) AS quantity,
            (IFNULL(pod.unit_weight, 0) * IFNULL(pod.required_qty, 0)) AS total_weight
        FROM `tabFT Project` p
        LEFT JOIN `tabFT Add Drawing` ad ON ad.project_number = p.name
        LEFT JOIN `tabFT Po Drawing`  pod ON pod.drawing_number = ad.name
        WHERE 1=1 {conditions}
        ORDER BY CAST(pod.po_serial_no AS UNSIGNED) ASC
    """

    # ✅ Sirf pure data rows — koi TOTAL row nahi
    raw_data = frappe.db.sql(query, values, as_dict=1) or []

    return columns, raw_data, None, None, None


# ------------------ Cards — Directly Doctype se (table se bilkul alag)
@frappe.whitelist()
def get_summary_cards(filters=None):
    import json

    if filters:
        filters = json.loads(filters) if isinstance(filters, str) else filters
    else:
        filters = {}

    # ---------- FT Project doctype se ----------
    proj_cond = "WHERE 1=1"
    proj_vals = {}

    if filters.get("project_number"):
        proj_cond += " AND name IN %(project_number)s"
        proj_vals["project_number"] = tuple(filters["project_number"])

    if filters.get("is_active"):
        proj_cond += " AND is_active = 1"

    total_projects = frappe.db.sql(
        f"SELECT COUNT(*) FROM `tabFT Project` {proj_cond}", proj_vals
    )[0][0] or 0

    project_total_weight = frappe.db.sql(
        f"SELECT IFNULL(SUM(total_weight), 0) FROM `tabFT Project` {proj_cond}", proj_vals
    )[0][0] or 0

    # ---------- FT Add Drawing + FT Po Drawing se ----------
    join_cond = "WHERE 1=1"
    join_vals = {}

    if filters.get("project_number"):
        join_cond += " AND p.name IN %(project_number)s"
        join_vals["project_number"] = tuple(filters["project_number"])

    if filters.get("drawing_number"):
        join_cond += " AND ad.drawing_number IN %(drawing_number)s"
        join_vals["drawing_number"] = tuple(filters["drawing_number"])

    if filters.get("po_serial_no"):
        join_cond += " AND pod.po_serial_no IN %(po_serial_no)s"
        join_vals["po_serial_no"] = tuple(str(x) for x in filters["po_serial_no"])

    if filters.get("is_active"):
        join_cond += " AND p.is_active = 1"

    po_result = frappe.db.sql(f"""
        SELECT
            COUNT(DISTINCT ad.name)                             AS total_drawings,
            COUNT(pod.name)                                     AS total_line_items,
            IFNULL(SUM(pod.unit_weight * pod.required_qty), 0) AS total_weight_line_items
        FROM `tabFT Project` p
        LEFT JOIN `tabFT Add Drawing` ad  ON ad.project_number  = p.name
        LEFT JOIN `tabFT Po Drawing`  pod ON pod.drawing_number = ad.name
        {join_cond}
    """, join_vals, as_dict=1)

    po_data           = po_result[0] if po_result else {}
    total_drawings    = po_data.get("total_drawings",        0) or 0
    total_line_items  = po_data.get("total_line_items",      0) or 0
    total_weight_line = po_data.get("total_weight_line_items", 0) or 0
    drawing_balance   = float(project_total_weight) - float(total_weight_line)

    return {
        "total_projects":          int(total_projects),
        "project_total_weight":    float(project_total_weight),
        "total_line_items":        int(total_line_items),
        "total_weight_line_items": float(total_weight_line),
        "total_drawings":          int(total_drawings),
        "drawing_balance":         float(drawing_balance),
    }


# ------------------ Drawing Number filter
@frappe.whitelist()
def get_drawing_numbers(project_number=None, txt=None):
    import json

    filters = {}
    if project_number:
        projects = json.loads(project_number) if isinstance(project_number, str) else project_number
        if projects:
            filters["project_number"] = ["in", projects]

    results = frappe.db.get_all("FT Add Drawing", filters=filters, fields=["drawing_number"], limit=0)

    seen, unique = set(), []
    for r in results:
        val = r.get("drawing_number")
        if val and val not in seen:
            seen.add(val)
            unique.append({"value": val, "label": val, "description": ""})

    if txt:
        tl = txt.lower()
        unique = [u for u in unique if tl in u["label"].lower()]

    return unique


# ------------------------- PO Serial No filter
@frappe.whitelist()
def get_po_serial_numbers(project_number=None, drawing_number=None, txt=None):
    import json

    project_list = json.loads(project_number) if isinstance(project_number, str) else (project_number or [])
    drawing_list = json.loads(drawing_number) if isinstance(drawing_number, str) else (drawing_number or [])

    cond, vals = "WHERE 1=1", {}

    if project_list:
        cond += " AND pod.project_number IN %(project_list)s"
        vals["project_list"] = tuple(project_list)

    if drawing_list:
        cond += " AND pod.drawing_number IN (SELECT name FROM `tabFT Add Drawing` WHERE drawing_number IN %(drawing_list)s)"
        vals["drawing_list"] = tuple(drawing_list)

    rows = frappe.db.sql(
        f"SELECT DISTINCT pod.po_serial_no FROM `tabFT Po Drawing` pod {cond} "
        f"ORDER BY CAST(pod.po_serial_no AS UNSIGNED) ASC",
        vals, as_dict=True
    )

    unique = []
    for r in rows:
        val = r.get("po_serial_no")
        if val is not None:
            s = str(val)
            unique.append({"value": s, "label": s, "description": ""})

    if txt:
        tl = txt.lower()
        unique = [u for u in unique if tl in u["label"].lower()]

    return unique


# ------------------------- Excel Export
@frappe.whitelist()
def download_drawing_excel(filters):
    import openpyxl
    from openpyxl.styles import Font, PatternFill, Border, Side, Alignment
    from io import BytesIO

    filters = frappe.parse_json(filters)

    conditions, values = "", {}

    if filters.get("project_number"):
        conditions += " AND p.name IN %(project_number)s"
        values["project_number"] = tuple(filters.get("project_number"))

    if filters.get("drawing_number"):
        conditions += " AND pod.drawing_number IN %(drawing_number)s"
        values["drawing_number"] = tuple(filters.get("drawing_number"))

    data = frappe.db.sql(f"""
        SELECT p.name AS project,
               IFNULL(pod.drawing_number, '') AS drawing_number,
               IFNULL(pod.po_serial_no, '')   AS po_serial_no,
               IFNULL(pod.unit_weight, 0)     AS unit_weight,
               IFNULL(pod.required_qty, 0)    AS required_qty,
               IFNULL(pod.total_weight, 0)    AS total_weight
        FROM `tabFT Project` p
        LEFT JOIN `tabFT Po Drawing` pod ON p.name = pod.project_number
        WHERE 1=1 {conditions}
        ORDER BY p.name
    """, values, as_dict=True)

    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "PO Drawing Data"

    ws.merge_cells('A1:F1')
    c = ws['A1']
    c.value = "PO Drawing"
    c.font  = Font(size=16, bold=True)
    c.alignment = Alignment(horizontal="center", vertical="center")

    hdrs  = ["Project", "Drawing Number", "PO Serial No", "Unit Weight", "Required Qty", "Total Weight"]
    hfont = Font(bold=True, size=13, color="FFFFFF")
    hfill = PatternFill(start_color="2F75B5", end_color="2F75B5", fill_type="solid")
    dfont = Font(size=12)
    tfont = Font(bold=True, size=13, color="FFFFFF")
    tfill = PatternFill(start_color="2F75B5", end_color="2F75B5", fill_type="solid")
    bdr   = Border(left=Side(style="thin"), right=Side(style="thin"),
                   top=Side(style="thin"),  bottom=Side(style="thin"))
    ctr   = Alignment(horizontal="center", vertical="center")

    for col, h in enumerate(hdrs, 1):
        cell = ws.cell(row=2, column=col, value=h)
        cell.font = hfont; cell.fill = hfill; cell.border = bdr; cell.alignment = ctr

    rn = 3; uwt = rqt = gt = 0

    for d in data:
        uw = d.get("unit_weight") or 0
        rq = d.get("required_qty") or 0
        tw = d.get("total_weight") or 0
        uwt += uw; rqt += rq; gt += tw

        for col, val in enumerate([d.get("project"), d.get("drawing_number"),
                                    d.get("po_serial_no"), uw, rq, tw], 1):
            cell = ws.cell(row=rn, column=col, value=val)
            cell.font = dfont; cell.border = bdr; cell.alignment = ctr
            if col in [4, 6]: cell.number_format = '0.000'
        rn += 1

    # Total row Excel mein rakhni hai (export ke liye zaroori)
    for col in range(1, 7):
        ws.cell(row=rn, column=col).fill   = tfill
        ws.cell(row=rn, column=col).border = bdr

    tl = ws.cell(row=rn, column=1, value="Total")
    tl.font = tfont
    tl.alignment = Alignment(horizontal="left", vertical="center")

    for col, val, fmt in [(4, uwt, '0.000'), (5, rqt, None), (6, gt, '0.000')]:
        c = ws.cell(row=rn, column=col, value=val)
        c.font = tfont; c.fill = tfill; c.alignment = ctr
        if fmt: c.number_format = fmt

    ws.column_dimensions["A"].width = 20; ws.column_dimensions["B"].width = 22
    ws.column_dimensions["C"].width = 18; ws.column_dimensions["D"].width = 15
    ws.column_dimensions["E"].width = 15; ws.column_dimensions["F"].width = 18

    fs = BytesIO()
    wb.save(fs); fs.seek(0)

    frappe.response['filename']    = "Drawing_Report.xlsx"
    frappe.response['filecontent'] = fs.getvalue()
    frappe.response['type']        = 'binary'

File: fabtrk/fabtrk/report/ft_drawings/ft_drawings.js
// Copyright (c) 2026, UpGo Technologies and contributors
// For license information, please see license.txt

frappe.query_reports["FT Drawings"] = {

	onload(report) {
		frappe.query_report.set_filter_value("project_number", []);
		frappe.query_report.set_filter_value("drawing_number", []);
		frappe.query_report.set_filter_value("po_serial_no", []);
		frappe.query_report.set_filter_value("is_active", 1);

		report.page.add_inner_button("Drawing Excel", function () {
			let filters = report.get_values();
			let url = "/api/method/fabtrk.fabtrk.report.ft_drawings.ft_drawings.download_drawing_excel"
				+ "?filters=" + encodeURIComponent(JSON.stringify(filters));
			window.location.href = url;
		});

		setTimeout(() => frappe.query_report.refresh(), 100);
	},

	// ✅ after_datatable_render — ft_drawing_part.js wala pattern follow kar raha hai
	after_datatable_render(report) {
		render_cards_from_api();
		attach_sticky_total_footer(report);   
	},

	filters: [

		// ---------- PROJECT ----------
		{
			fieldname: "project_number",
			label: "Project Number",
			fieldtype: "MultiSelectList",
			get_data: function (txt) {
				let is_active = frappe.query_report.get_filter_value("is_active");
				let filters = {};
				if (is_active) filters.is_active = 1;
				return frappe.call({
					method: "frappe.client.get_list",
					args: { doctype: "FT Project", filters: filters, fields: ["name", "description"], limit: 0 }
				}).then(r => {
					let result = (r.message || []).map(d => ({
						value: d.name, label: d.name, description: d.description || ""
					}));
					if (txt) result = result.filter(i => i.label.toLowerCase().includes(txt.toLowerCase()));
					return result;
				});
			},
			on_change() {
				frappe.query_report.set_filter_value("drawing_number", []);
				frappe.query_report.set_filter_value("po_serial_no", []);
				frappe.query_report.refresh();
				setTimeout(() => frappe.query_report.refresh(), 100);
			}
		},

		// ---------- DRAWING ----------
		{
			fieldname: "drawing_number",
			label: "Drawing Number",
			fieldtype: "MultiSelectList",
			get_data: function (txt) {
				let projects = frappe.query_report.get_filter_value("project_number");
				return frappe.call({
					method: "fabtrk.fabtrk.report.ft_drawings.ft_drawings.get_drawing_numbers",
					args: { project_number: JSON.stringify(projects || []), txt: txt || "" }
				}).then(r => r.message || []);
			},
			on_change() {
				frappe.query_report.set_filter_value("po_serial_no", []);
				frappe.query_report.refresh();
			}
		},

		// ---------- PO SERIAL NO ----------
		{
			fieldname: "po_serial_no",
			label: "PO Serial No",
			fieldtype: "MultiSelectList",
			get_data: function (txt) {
				let projects = frappe.query_report.get_filter_value("project_number");
				let drawings = frappe.query_report.get_filter_value("drawing_number");
				return frappe.call({
					method: "fabtrk.fabtrk.report.ft_drawings.ft_drawings.get_po_serial_numbers",
					args: {
						project_number: JSON.stringify(projects || []),
						drawing_number: JSON.stringify(drawings || []),
						txt: txt || ""
					}
				}).then(r => r.message || []);
			},
			on_change() { frappe.query_report.refresh(); }
		},

		// ---------- IS ACTIVE ----------
		{
			fieldname: "is_active",
			label: "Is Active",
			fieldtype: "Check",
			default: 1,
			on_change() {
				frappe.query_report.set_filter_value("project_number", []);
				frappe.query_report.set_filter_value("drawing_number", []);
				frappe.query_report.set_filter_value("po_serial_no", []);
				frappe.query_report.refresh();
			}
		}
	],

	formatter: function (value, row, column, data, default_formatter) {
		return default_formatter(value, row, column, data);
	}
};


// ------------------ STICKY TOTAL FOOTER 
function attach_sticky_total_footer(report) {
	// Pehle purana footer hata do
	$(report.wrapper).find("#ft-sticky-total-footer").remove();

	let $wrapper  = $(report.wrapper);
	let $dt_body  = $wrapper.find(".dt-scrollable");
	if (!$dt_body.length) return;

	// ── Data aur columns frappe se lo
	let data    = frappe.query_report.data    || [];
	let columns = frappe.query_report.columns || [];

	// ── FT Drawings ke columns ke totals
	// Sirf numeric columns ka total: unit_weight, quantity, total_weight
	let totals = {
		unit_weight:  0,
		quantity:     0,
		total_weight: 0,
	};

	data.forEach(function (d) {
		// TOTAL naam ki row already nahi hai (Python mein nahi daali)
		// Phir bhi safe raho
		if (d.project_name === "TOTAL") return;
		totals.unit_weight  += parseFloat(d.unit_weight  || 0);
		totals.quantity     += parseFloat(d.quantity      || 0);
		totals.total_weight += parseFloat(d.total_weight  || 0);
	});

	// ── Header cells se exact widths lo (ft_drawing_part.js wala same approach)
	let col_widths = [];
	$wrapper.find(".dt-header .dt-cell").each(function () {
		let w = $(this).outerWidth();
		col_widths.push(w > 0 ? w : 100);
	});

	if (!col_widths.length) return;

	// ── Format helper
	function fmt(num, dec) {
		dec = dec !== undefined ? dec : 3;
		return Number(num).toLocaleString("en-IN", {
			minimumFractionDigits: dec,
			maximumFractionDigits: dec
		});
	}

	// ── Serial number column (col index 0) — blank
	let cells_html = `
		<div style="
			width:${col_widths[0]}px;
			min-width:${col_widths[0]}px;
			max-width:${col_widths[0]}px;
			display:inline-flex;
			align-items:center;
			justify-content:center;
			padding:0 4px;
			border-right:1px solid #d1d8dd;
			flex-shrink:0;
			box-sizing:border-box;
			height:40px;
			background:#f3f3f3;
		"></div>`;

	// ── Data columns — columns array se iterate karo
	columns.forEach(function (col, i) {
		let w  = col_widths[i + 1] || 120;   // i+1 kyunki 0 serial number ka tha
		let fn = col.fieldname;
		let val = "";
		let align = "center";

		if (fn === "project_name") {
			// ✅ TOTAL label — left aligned, bold
			val   = `<span style="font-weight:900; font-size:13px; color:#000;">TOTAL</span>`;
			align = "left";

		} else if (fn === "unit_weight") {
			val = fmt(totals.unit_weight, 3);

		} else if (fn === "quantity") {
			val = fmt(totals.quantity, 0);

		} else if (fn === "total_weight") {
			val = fmt(totals.total_weight, 3);

		}
		// drawing_number, po_serial_no — blank rehenge

		let is_last    = (i === columns.length - 1);
		let border_right = is_last ? "none" : "1px solid #d1d8dd";

		cells_html += `
			<div style="
				width:${w}px;
				min-width:${w}px;
				max-width:${w}px;
				display:inline-flex;
				align-items:center;
				justify-content:${align === "left" ? "flex-start" : "center"};
				padding:0 8px;
				border-right:${border_right};
				flex-shrink:0;
				box-sizing:border-box;
				height:40px;
				font-weight:700;
				font-size:13px;
				color:#000;
				background:#f3f3f3;
				white-space:nowrap;
				overflow:hidden;
			">${val}</div>`;
	});

	// ── Footer DOM inject karo — dt-scrollable ke BAAD (ft_drawing_part.js wala same)
	let $footer = $(`
		<div id="ft-sticky-total-footer">
			<div id="ft-footer-inner" style="
				display: inline-flex;
				flex-wrap: nowrap;
				transform: translateX(0px);
				height: 40px;
			">${cells_html}</div>
		</div>
	`);

	$dt_body.after($footer);

	// ── Horizontal scroll sync — dt-scrollable scroll hone par footer bhi saath move kare
	$dt_body.off("scroll.ft_drawings_footer").on("scroll.ft_drawings_footer", function () {
		$("#ft-footer-inner").css("transform", `translateX(-${this.scrollLeft}px)`);
	});
}

// ------------------------ CARDS — Directly Doctype API se (table se bilkul alag)
function render_cards_from_api() {
	let filters = frappe.query_report.get_values() || {};

	frappe.call({
		method: "fabtrk.fabtrk.report.ft_drawings.ft_drawings.get_summary_cards",
		args: { filters: JSON.stringify(filters) },
		callback: function (r) {
			if (!r.message) return;
			render_cards_html(r.message);
		}
	});
}

function render_cards_html(data) {
	$(".ft-custom-summary-2").remove();

	function fmt(num) {
		return Number(num).toLocaleString("en-IN", {
			minimumFractionDigits: 3,
			maximumFractionDigits: 3
		});
	}

	let html = `
	<div class="ft-custom-summary-2">
		<div class="ft-card-2">
			<p class="ft-card-label-2">Total Projects</p>
			<span class="ft-card-value-2">${data.total_projects}</span>
			<p class="ft-card-label-2" style="margin-top:12px;">Total Weight as per Project (Kg)</p>
			<span class="ft-card-value-2">${fmt(data.project_total_weight)}</span>
		</div>
		<div class="ft-card-2">
			<p class="ft-card-label-2">Project Total No of Line Items</p>
			<span class="ft-card-value-2">${data.total_line_items}</span>
			<p class="ft-card-label-2" style="margin-top:12px;">Total Weight of Line Items Assign to Project (Kg)</p>
			<span class="ft-card-value-2">${fmt(data.total_weight_line_items)}</span>
		</div>
		<div class="ft-card-2">
			<p class="ft-card-label-2">Total No of Unique Drawings</p>
			<span class="ft-card-value-2">${data.total_drawings}</span>
		</div>
		<div class="ft-card-2">
			<p class="ft-card-label-2">Drawing Balance for Customer (Kg)</p>
			<span class="ft-card-value-2">${fmt(data.drawing_balance)}</span>
		</div>
	</div>`;

	$(".report-wrapper").prepend(html);
}

$(`<style>

/* ===== STICKY FOOTER — ft_drawing_part.js se same pattern ===== */
#ft-sticky-total-footer {
	position: sticky !important;
	bottom: 0 !important;
	z-index: 50 !important;
	background-color: #f3f3f3 !important;
	border-top: 3px solid #d1d8dd !important;
	width: 100% !important;
	display: flex !important;
	overflow: hidden !important;
	height: 40px !important;
}

#ft-sticky-total-footer > div {
	background-color: #f3f3f3 !important;
}

#ft-footer-inner {
	will-change: transform;
}


/* ===== CARDS ===== */
.ft-custom-summary-2 {
	display: grid;
	grid-template-columns: repeat(4, 1fr);
	gap: 20px;
	width: 100%;
	padding: 20px 0;
}
@media (max-width: 768px) {
	.ft-custom-summary-2 { grid-template-columns: repeat(1, 1fr) !important; }
	.datatable .dt-scrollable { overflow-x: auto !important; -webkit-overflow-scrolling: touch; }
}
.ft-card-2 {
	background: #fff;
	padding: 20px;
	border-radius: 10px;
	text-align: center;
	border: 2px solid #eef0f4;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	transition: box-shadow 0.2s ease;
}
.ft-card-label-2 { font-size: 13px; color: #6c757d; margin: 0 0 4px 0; text-align: center; width: 100%; }
.ft-card-value-2 { font-size: 18px; font-weight: 700; color: #1a1a2e; display: block; text-align: center; width: 100%; }


/* ===== TABLE ===== */
.report-wrapper, .datatable, .datatable-container { width: 100% !important; }
.datatable table { width: 100% !important; table-layout: auto !important; }
.datatable .dt-cell { white-space: nowrap !important; }
.datatable .dt-cell__content {
	text-align: center !important;
	justify-content: center !important;
	overflow: visible !important;
	text-overflow: unset !important;
}
.datatable .dt-row { width: 100% !important; }
.datatable .dt-scrollable { overflow-x: auto !important; }

</style>`).appendTo("head");







